src/connection.c

```diff
@@ -1,453 +1,214 @@
- *    allows various context and configuration setting to be handled before
- *    establishing the actual connection.
- * 2. The caller may register/unregister logical read/write handlers to be
- *    called when the connection has data to read from/can accept writes.
- *    These logical handlers may or may not correspond to actual AE events,
- *    depending on the implementation (for TCP they are; for TLS they aren't).
- */
+static ConnectionType *connTypes[CONN_TYPE_MAX];
 
-ConnectionType CT_Socket;
+int connTypeRegister(ConnectionType *ct) {
+    const char *typename = ct->get_type(NULL);
+    ConnectionType *tmpct;
+    int type;
 
-/* When a connection is created we must know its type already, but the
- * underlying socket may or may not exist:
- *
- * - For accepted connections, it exists as we do not model the listen/accept
- *   part; So caller calls connCreateSocket() followed by connAccept().
- * - For outgoing connections, the socket is created by the connection module
- *   itself; So caller calls connCreateSocket() followed by connConnect(),
- *   which registers a connect callback that fires on connected/error state
- *   (and after any transport level handshake was done).
- *
- * NOTE: An earlier version relied on connections being part of other structs
- * and not independently allocated. This could lead to further optimizations
- * like using container_of(), etc.  However it was discontinued in favor of
- * this approach for these reasons:
- *
- * 1. In some cases conns are created/handled outside the context of the
- * containing struct, in which case it gets a bit awkward to copy them.
- * 2. Future implementations may wish to allocate arbitrary data for the
- * connection.
- * 3. The container_of() approach is anyway risky because connections may
- * be embedded in different structs, not just client.
- */
-
-connection *connCreateSocket() {
-    connection *conn = zcalloc(sizeof(connection));
-    conn->type = &CT_Socket;
-    conn->fd = -1;
-
-    return conn;
-}
+    /* find an empty slot to store the new connection type */
+    for (type = 0; type < CONN_TYPE_MAX; type++) {
+        tmpct = connTypes[type];
+        if (!tmpct)
+            break;
 
-/* Create a new socket-type connection that is already associated with
- * an accepted connection.
- *
- * The socket is not ready for I/O until connAccept() was called and
- * invoked the connection-level accept handler.
- *
- * Callers should use connGetState() and verify the created connection
- * is not in an error state (which is not possible for a socket connection,
- * but could but possible with other protocols).
- */
-connection *connCreateAcceptedSocket(int fd) {
-    connection *conn = connCreateSocket();
-    conn->fd = fd;
-    conn->state = CONN_STATE_ACCEPTING;
-    return conn;
-}
+        /* ignore case, we really don't care "tls"/"TLS" */
+        if (!strcasecmp(typename, tmpct->get_type(NULL)))
```

Comment:
Any reason not to simply return an error here, and panic only if we fail trying to register the built-in types?

Reply:
> Any reason not to simply return an error here, and panic only if we fail trying to register the built-in types?

Connection type is allowed to be loaded during Redis startup only, at this stage, reporting unexpected connection types seems obvious.
For example, `./src/redis-server --loadmodule src/redis-tls.so  --loadmodule src/redis-foo.so` tries to load 2 modules, unfortunately they have the same connection name `TLS`, if we don't panic here, we don't know which one is in use.

Reply:
I think Yossi meant that the panic (or rather `exit`), is handled outside.
i.e. we'll return an error here, it'll cause the module to fail loading, and then moduleLoadFromQueue will fail.
it would mean the user doesn't know the reason was a conflicting name, unless we add a log print either here, or in the module when that call returns with an error.